File: backend/app/schemas/device.py
from datetime import datetime
from ipaddress import ip_address

from pydantic import BaseModel, Field, field_validator

from app.models.device import DeviceStatus, DeviceType, DeviceVendor


class DeviceBase(BaseModel):
    device_name: str = Field(min_length=1, max_length=255)
    hostname: str | None = Field(default=None, max_length=255)
    vendor: DeviceVendor
    device_type: DeviceType
    ip_address: str = Field(min_length=2, max_length=45)
    location_id: str | None = None
    parent_device_id: str | None = None
    status: DeviceStatus = DeviceStatus.UNKNOWN
    description: str | None = None

    @field_validator("ip_address")
    @classmethod
    def validate_ip_address(cls, value: str) -> str:
        try:
            return str(ip_address(value))
        except ValueError as exc:
            raise ValueError("ip_address must be a valid IPv4 or IPv6 address") from exc


class DeviceCreate(DeviceBase):
    pass


class DeviceUpdate(BaseModel):
    device_name: str | None = Field(default=None, min_length=1, max_length=255)
    hostname: str | None = Field(default=None, max_length=255)
    vendor: DeviceVendor | None = None
    device_type: DeviceType | None = None
    ip_address: str | None = Field(default=None, min_length=2, max_length=45)
    location_id: str | None = None
    parent_device_id: str | None = None
    status: DeviceStatus | None = None
    description: str | None = None

    @field_validator("ip_address")
    @classmethod
    def validate_ip_address(cls, value: str | None) -> str | None:
        if value is None:
            return value
        try:
            return str(ip_address(value))
        except ValueError as exc:
            raise ValueError("ip_address must be a valid IPv4 or IPv6 address") from exc


class DeviceResponse(DeviceBase):
    id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DeviceListResponse(BaseModel):
    items: list[DeviceResponse]
    total: int
    page: int
    page_size: int


class DeviceBulkImportRequest(BaseModel):
    devices: list[DeviceCreate] = Field(min_length=1, max_length=500)


class DeviceBulkImportResponse(BaseModel):
    created: list[DeviceResponse]
    failed: list[dict[str, str]]

File: frontend/lib/src/features/devices/domain/device.dart
class Device {
  const Device({
    required this.id,
    required this.deviceName,
    required this.vendor,
    required this.deviceType,
    required this.ipAddress,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.hostname,
    this.locationId,
    this.parentDeviceId,
    this.description,
  });

  final String id;
  final String deviceName;
  final String? hostname;
  final String vendor;
  final String deviceType;
  final String ipAddress;
  final String? locationId;
  final String? parentDeviceId;
  final String status;
  final String? description;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory Device.fromJson(Map<String, dynamic> json) {
    return Device(
      id: json['id'] as String,
      deviceName: json['device_name'] as String,
      hostname: json['hostname'] as String?,
      vendor: json['vendor'] as String,
      deviceType: json['device_type'] as String,
      ipAddress: json['ip_address'] as String,
      locationId: json['location_id'] as String?,
      parentDeviceId: json['parent_device_id'] as String?,
      status: json['status'] as String,
      description: json['description'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

class DeviceListResult {
  const DeviceListResult({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  final List<Device> items;
  final int total;
  final int page;
  final int pageSize;

  factory DeviceListResult.fromJson(Map<String, dynamic> json) {
    return DeviceListResult(
      items: (json['items'] as List<dynamic>)
          .map((item) => Device.fromJson(item as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
    );
  }
}
